Teach TSan about TaskGroups

yln opened the pull request:
`TaskGroup::offer(completedTask)`:
* Called from `AsyncTask::completeFuture()` which already takes care of
  `release(completedTask)`.  No additional edge is required if no task
  is waiting and completed task is stored so it can be retrieved by
  `group.poll()`.
* If group has waiting task, it will be dequeued and scheduled, add
  `acquire(waitingTask)`.

`TaskGroupImpl::poll(waitingTask)`:
* If no pending tasks, do nothing.
* If returning finished task, `acquire(finishedTask)`.
* If enqueuing waiting task, `release(waitingTask)`.

Note: `release()` should go before, and `acquire()` after the annotated
      synchronization operation.

Radar-Id: rdar://75910632

Commit:
Teach TSan about TaskGroups

`TaskGroup::offer(completedTask)`:
* Called from `AsyncTask::completeFuture()` which already takes care of
  `release(completedTask)`.  No additional edge is required if no task
  is waiting and completed task is stored so it can be retrieved by
  `group.poll()`.
* If group has waiting task, it will be dequeued and scheduled, add
  `acquire(waitingTask)`.

`TaskGroupImpl::poll(waitingTask)`:
* If no pending tasks, do nothing.
* If returning finished task, `acquire(finishedTask)`.
* If enqueuing waiting task, `release(waitingTask)`.

Note: `release()` should go before, and `acquire()` after the annotated
      synchronization operation.

Radar-Id: rdar://75910632

Commit:
Add test for TSan + TaskGroups

yln commented:
Although I was creating more and more complex setups, I was still unable to create a test that "goes red" (TSan producing a FP) to proof that these annotations are needed.  So I simplified the test again to make it easier to understand.  @ktoso Do you have ideas/suggestions on how a test that "correctly goes red" could look like?

ktoso commented:
Okey, I'm a bit green in how TSAN works/expects the order of the release/acquire it somehow reads reversed to me, but maybe I'm just confused?

It looks good and your writeup is correct I think, to double check though: there's both orders which make sense in a group:

- offer runs, no waiter exists -> `release(completedTask AKA readyTask)`
- poll runs, ready task exists -> `acquire(readyTask)`

and

- poll runs, no ready task, enqueue waiter -> `release(waitingTask)`
- offer runs, waiter exists, run the waiter -> `acquire(waitingTask)`

So is it generally this feels like "when we're about to run a task, we acquire it" and when we're "done" with a task we release it", did I get that right?

The edges seem right then and I think you identified all of them - it can only be "ready + no waiter" "ready + waiter" and "waiter but nothing to wait on anyway (noop)" as you identified correctly up there.

yln commented:
> Okey, I'm a bit green in how TSAN works/expects the order of the release/acquire it somehow reads reversed to me, but maybe I'm just confused?

I always have to remind myself:
```
 T1             T2
 tsan_release(x)
 unlock(x)
          -->
             lock(x)
             tsan_acquire(x)
```
In our case `unlock()` is "task enqueued/suspended". `lock()` is "task dequeued/resumed".

> 
> It looks good and your writeup is correct I think, to double check though: there's both orders which make sense in a group:
> 
> * offer runs, no waiter exists -> `release(completedTask AKA readyTask)`

This is always done (sometimes unnecessarily w.r.t. task groups) in `AsyncTask::completeFuture()`.

> * poll runs, ready task exists -> `acquire(readyTask)`

Added by this PR.

> 
> and
> 
> * poll runs, no ready task, enqueue waiter -> `release(waitingTask)`
> * offer runs, waiter exists, run the waiter -> `acquire(waitingTask)`

Added by this PR.

> 
> So is it generally this feels like "when we're about to run a task, we acquire it" and when we're "done" with a task we release it", did I get that right?

Yes!

yln commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -744,6 +746,7 @@ PollResult TaskGroupImpl::poll(AsyncTask *waitingTask) {
           result.storage = futureFragment->getStoragePtr();
           assert(result.retainedTask && "polled a task, it must be not null");
+          _swift_tsan_acquire(static_cast<Job *>(result.retainedTask));
           mutex.unlock(); // TODO: remove fragment lock, and use status for synchronization
```
Konrad also pointed out that the current implementation uses a lock (pending for removal), which of course adds additional edges for TSan.  This may explain my futile attempts of producing a test that "goes red".

yln commented:
@swift-ci Please test

yln commented:
Build on Linux failed with Jenkins error:
```
[639/1291][ 49%][259.954s] Building CXX object lib/PrintAsObjC/CMakeFiles/swiftPrintAsObjC.dir/DeclAndTypePrinter.cpp.o
FATAL: command execution failed
java.io.EOFException
	at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2759)
	at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3234)
	at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:913)
	at java.io.ObjectInputStream.<init>(ObjectInputStream.java:375)
	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:49)
	at hudson.remoting.Command.readFrom(Command.java:142)
	at hudson.remoting.Command.readFrom(Command.java:128)
	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)
	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)
Caused: java.io.IOException: Unexpected termination of the channel
```

yln commented:
@swift-ci Please test Linux

yln commented:
Linux build again failed during compilation with a Jenkins/Java exception.

yln commented:
@swift-ci Please test Linux

ktoso commented:
@swift-ci please smoke test linux

ktoso commented:
Would we want to merge this? @yln

yln commented:
@swift-ci Please smoke test and merge

yln commented:
```
14:26:18 Failed Tests (1):
14:26:18   Swift(macosx-x86_64) :: Sanitizers/tsan/async_taskgroup_next.swift

Compilation fails:
async_taskgroup_next.swift:12:11: error: 'Task' is only available in macOS 9999 or newer
```

I guess we have to annotate the test like this:
https://github.com/apple/swift/commit/1173b737aa3422137eb85079b7768a6932269158

yln commented:
Updated test to reflect TaskGroup API changes and annotated with availability.

yln commented:
@swift-ci Please smoke test and merge
